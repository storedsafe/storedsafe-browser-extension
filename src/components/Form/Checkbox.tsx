import * as React from 'react';
import './Checkbox.scss';

type CheckboxProps = React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
export const Checkbox: React.SFC<CheckboxProps> = (props: CheckboxProps) => (
  <React.Fragment>
    <div style={{ display: 'inline-block' }} className="checkbox">
      <input type="checkbox" {...props} />
      <div className="custom-checkbox" />
    </div>
  </React.Fragment>
);
