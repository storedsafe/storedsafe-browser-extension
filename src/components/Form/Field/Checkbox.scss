@import '../../../variables';

.field-checkbox {
  $size: 20px;
  $border-size: 2px;
  $padding: 2px;
  $inner-size: $size - ($border-size * 2) - ($padding * 2);

  .field-container {
    display: flex;
    flex-direction: column;
    border: 0;
  }

  label {
    display: flex;
    align-items: center;
  }

  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  .custom-checkbox {
    padding: $padding;
    position: relative;
    border: $border-size solid $border;
    border-radius: $border-radius;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset;
    width: $size;
    height: $size;
    margin-right: 0.5em;
  }

  .custom-checkbox::before {
    display: none;
    position: absolute;
    content: '';
    width: $inner-size;
    height: $inner-size;
    background-color: $accent;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1);
    border-radius: $border-radius;
  }

  input:checked ~ .custom-checkbox::before {
    display: block;
  }

  input:hover ~ .custom-checkbox,
  input:active ~ .custom-checkbox,
  input:focus ~ .custom-checkbox {
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset, 0 0 3px 0 rgba(0, 0, 0, 0.1);
  }

  input:hover ~ .custom-checkbox {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  input:focus ~ .custom-checkbox {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  input:active ~ .custom-checkbox {
    border-color: $border-active;
    background-color: $bg-active;
  }
}
