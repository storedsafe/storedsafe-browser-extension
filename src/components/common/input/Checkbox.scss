@import '../../../styles/variables.scss';

.checkbox {
  $size: 20px;
  $border-size: 1px;
  $padding: 2px;
  $inner-size: $size - ($border-size * 2) - ($padding * 2);

  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  .custom-checkbox {
    padding: $padding;
    position: relative;
    border: $border-size solid $color-border;
    border-radius: 2px;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset;
    width: $size;
    height: $size;
    cursor: pointer;
    box-sizing: border-box;
  }

  .custom-checkbox::before {
    display: none;
    position: absolute;
    content: '';
    width: $inner-size;
    height: $inner-size;
    background-color: $color-accent-0;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1);
    border-radius: 2px;
  }

  input:checked ~ .custom-checkbox::before {
    display: block;
  }

  input:hover ~ .custom-checkbox,
  input:active ~ .custom-checkbox,
  input:focus ~ .custom-checkbox {
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset, 0 0 3px 0 rgba(0, 0, 0, 0.1);
  }

  input:hover ~ .custom-checkbox {
    border-color: $color-border-hover;
    background-color: $color-bg-hover;
  }

  input:focus ~ .custom-checkbox {
    border-color: $color-border-focus;
    background-color: $color-bg-hover;
  }

  input:active ~ .custom-checkbox {
    border-color: $color-border;
    background-color: $color-bg-active;
  }
}
