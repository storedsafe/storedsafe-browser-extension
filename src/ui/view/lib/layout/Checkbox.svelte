<script lang="ts" module>
  import type { HTMLAttributes } from "svelte/elements";

  export interface Props extends HTMLAttributes<HTMLInputElement> {
    size?: string;
    altered?: boolean;
    disabled?: boolean;
    checked?: boolean;
  }
</script>

<script lang="ts">
  let {
    size = "1em",
    altered = false,
    disabled = false,
    checked = $bindable(),
    ...restProps
  }: Props = $props();

  const style = [`--checkbox-size: ${size}`].join(";");
</script>

<span class="checkbox-container">
  <input
    bind:checked
    {disabled}
    type="checkbox"
    class="checkbox-hidden"
    {...restProps}
  />
  <span class="checkbox custom-input" class:altered {style}></span>
</span>

<style>
  .checkbox-container {
    --checkbox-size: 1em;
    position: relative;
    height: var(--checkbox-size);
    width: var(--checkbox-size);
  }

  .checkbox-hidden {
    position: absolute;
    opacity: 0;
    height: 0;
    width: 0;
  }

  .checkbox {
    position: absolute;
    font-size: var(--checkbox-size);
    top: 0;
    left: 0;
    height: var(--checkbox-size);
    width: var(--checkbox-size);
    cursor: pointer;
  }

  .checkbox-hidden:checked ~ .checkbox {
    background-color: var(--color-accent);
  }

  .checkbox:after {
    content: "";
    position: absolute;
    display: none;
  }

  .checkbox-hidden:checked ~ .checkbox:after {
    display: block;
  }

  .checkbox-hidden:disabled ~ .checkbox {
    background-color: var(--color-input-fg-disabled);
  }

  .checkbox-hidden:disabled:checked ~ .checkbox {
    background-color: var(--color-button-bg-disabled);
  }

  .checkbox:after {
    left: 30%;
    bottom: 17%;
    width: 40%;
    height: 80%;
    border: solid white;
    border-width: 0 0.15em 0.15em 0;
    box-sizing: border-box;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
</style>
