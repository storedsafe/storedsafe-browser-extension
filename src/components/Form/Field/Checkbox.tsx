import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Checkbox.scss';

type CheckboxProps = Pick<FieldProps, 'name' | 'value' | 'onChange' | 'onBlur' | 'label' | 'labelDir'>
export const Checkbox: React.SFC<CheckboxProps | FieldAttributes> = ({
  name,
  value,
  label,
  labelDir,
  onChange,
  onBlur,
  ...attributes
}: CheckboxProps) => (
  <div className="field field-checkbox">
    <div className="field-container">
      <label htmlFor={name} className="field-label">
        {labelDir === 'left' && (<span className="field-label">{label}</span>)}
        <input
          id={name}
          name={name}
          type="checkbox"
          checked={value as boolean}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            onChange && onChange(target.checked, target.name);
          }}
          onBlur={({ target }: React.FocusEvent<HTMLInputElement>): void => {
            onBlur && onBlur(target.checked, target.name);
          }}
          {...attributes}
        />
        <div className="custom-checkbox" />
        {labelDir === 'right' && (<span className="field-label">{label}</span>)}
      </label>
    </div>
  </div>
);

Checkbox.defaultProps = {
  value: false,
  labelDir: 'right',
};
