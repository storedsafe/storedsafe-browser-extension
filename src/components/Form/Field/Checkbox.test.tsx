import * as React from 'react';
import { Checkbox } from './Checkbox';
import { shallow, mount } from 'enzyme';

const onChange = jest.fn();
const onBlur = jest.fn();

beforeEach(() => {
  onChange.mockClear();
  onBlur.mockClear();
});

test('<Checkbox />', () => {
  const wrapper = shallow(<Checkbox name="test" label="Test" value={true} />);
  const checkbox = wrapper.find('input#test');
  expect(checkbox.props().type).toBe('checkbox');
  expect(checkbox.props().checked).toBe(true);
  expect(wrapper.find('.custom-checkbox').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('<Checkbox labelDir=left/>', () => {
  const wrapper = shallow(<Checkbox name="test" label="Test" value={true} labelDir="left" />);
  const checkbox = wrapper.find('input#test');
  expect(checkbox.props().type).toBe('checkbox');
  expect(checkbox.props().checked).toBe(true);
  expect(wrapper.find('.custom-checkbox').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('<Checkbox />, change', () => {
  const wrapper = mount(<Checkbox name="test" label="Test" value={false} onChange={onChange} />);
  wrapper.find('input#test').simulate(
    'change',
    { target: { name: 'test', checked: true } }
  );
  expect(onChange).toHaveBeenCalledWith(true, 'test');
});

test('<Checkbox />, blur', () => {
  const wrapper = mount(<Checkbox name="test" label="Test" value={false} onBlur={onBlur} />);
  wrapper.find('input#test').simulate(
    'blur',
    { target: { name: 'test', checked: true } }
  );
  expect(onBlur).toHaveBeenCalledWith(true, 'test');
});
