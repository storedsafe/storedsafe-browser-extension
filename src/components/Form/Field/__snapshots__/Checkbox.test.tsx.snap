// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> 1`] = `ShallowWrapper {}`;

exports[`<Checkbox labelDir=left/> 1`] = `ShallowWrapper {}`;
